### PR TITLE
Template refactor terraform

### PR DESCRIPTION
**Pull request by fredrikhgrelland**

### CODE REVIEW

**Comment by Neha-Sinha2305** [comment]
Though the matrix tests show up as green/passed, i checked that the env variables are not being overriden and all the combinations are running with the default values. Note the value for env variable **nomad_acl**.
Screenshot from 2020-08-14 09-28-43

**Comment by fredrikhgrelland** [comment]
> Though the matrix tests show up as green/passed, i checked that the env variables are not being overriden and all the combinations are running with the default values. Note the value for env variable **nomad_acl**.
> Screenshot from 2020-08-14 09-28-43

You are right. See bb12138 . Might fix it

**Comment by Neha-Sinha2305** [comment]
> > Though the matrix tests show up as green/passed, i checked that the env variables are not being overriden and all the combinations are running with the default values. Note the value for env variable **nomad_acl**.
> > Screenshot from 2020-08-14 09-28-43
> 
> You are right. See [bb12138](https://github.com/fredrikhgrelland/vagrant-hashistack/commit/bb121388df8e1c446c0362f8da4cd82c95d25eb6) . Might fix it

Will keep an eye on the next run. However, testing the tests makes less sense..the tests/test approach should be capable of detecting that the env variable values failed to get propagated or will end up getting false positives!

**Comment by fredrikhgrelland** [comment]
> > > Though the matrix tests show up as green/passed, i checked that the env variables are not being overriden and all the combinations are running with the default values. Note the value for env variable **nomad_acl**.
> > > Screenshot from 2020-08-14 09-28-43
> > 
> > 
> > You are right. See [bb12138](https://github.com/fredrikhgrelland/vagrant-hashistack/commit/bb121388df8e1c446c0362f8da4cd82c95d25eb6) . Might fix it
> 
> Will keep an eye on the next run. However, testing the tests makes less sense..the tests/test approach should be capable of detecting that the env variable values failed to get propagated or will end up getting false positives!

I agree completely. Any Ideas how to implement?

**Comment by Neha-Sinha2305** [comment]
Well, one of the ways would be to run specific tests for specific configurations or send the variables directly with the test run command. 
So, say in this case where we wanted to test **nomad_acl=true**, we could have triggered test by sending in this value via "run make test --nomad_acl=true".
Many testing frameworks provide the tagging capability where one could tag the tests and then trigger them based on the tags.

The basic issue with the current implementation is that we run the tests/checks by dynamically evaluating the env variable values. And these env variables get manipulated and populated by another piece of code and this piece of code can have bugs.

Also, If we could extract [this test](https://github.com/fredrikhgrelland/vagrant-hashistack/blob/6d77db88130c7885478a1f6382afcfcd948b2aa6/ansible/bootstrap.yml#L50) in a separate file **test_consul_acl_enabled_and_policy_deny** and remove the **when** block and run this test when consul_acl=true and consul_acl_default_policy=deny, we would get the correct feedback:
```
    - name: Verify Consul ACL deny policy - access deny without token
      uri:
        url: http://127.0.0.1:8500/v1/acl/tokens
        status_code: 403
      register: result
      until: result.status == 403
      retries: 5
      delay: 1
      ~~when: lookup('env', 'consul_acl') | bool and lookup('env', 'consul_acl_default_policy') == 'deny'~~
```

File: ansible/bootstrap.yml [hunk 1]
```diff
@@ -8,25 +8,31 @@
       - nomad
     token: master
   tasks:
+    - name: Check if .env_override is provided
+      debug:
+        msg: "We are running in test mode and .env_override is set"
+      failed_when: not lookup( 'env', 'local_test') and (not lookup( 'env', 'env_override') and not lookup( 'env', 'TF_VAR_env_override'))
```
Comment (Neha-Sinha2305) [inline ansible/bootstrap.yml:14]: Stupid question: Won't we achieve the same result from the below check? Wondering if we could reduce from 3 booleans to 2
```
failed_when: not lookup( 'env', 'local_test') and (not lookup( 'env', 'TF_VAR_env_override'))
```

Comment (fredrikhgrelland) [reply ansible/bootstrap.yml:14]: yes. Thanks!

Comment (fredrikhgrelland) [reply ansible/bootstrap.yml:14]: BTW: you can use the "suggest change"-functionality (+-) in the markdown editor to suggest directly

File: template/Makefile [hunk 2]
```diff
@@ -7,11 +7,24 @@ export PATH := $(shell pwd)/tmp:$(PATH)
 
 #### Development ####
 # start commands
-up: clean update-box
+dev: update-box
+	SSL_CERT_FILE=${SSL_CERT_FILE} CURL_CA_BUNDLE=${CURL_CA_BUNDLE} ANSIBLE_ARGS='--skip-tags "test"' vagrant up --provision
+
+up: update-box
+ifdef CI # CI is set in Github Actions
```
Comment (pdmthorsrud) [inline template/Makefile:14]: I might be missing something, but where in github actions is CI set?

File: template/dev/README.md [hunk 3]
```diff
@@ -5,4 +5,4 @@ Any configuration of the hashistack is best performed through terraform provider
 - [vault](https://www.terraform.io/docs/providers/vault/index.html)
 - [consul](https://www.terraform.io/docs/providers/consul/index.html)
 
-You put them in this directory and use [../ansible/playbook.yml](../ansible/playbook.yml) to init and run terraform.
\ No newline at end of file
+You put them in this directory and use [../ansible/playbook.yml](ansible/playbook.yml) to init and run terraform. `make up` will provision the box and run the playbooks inside the box.
```
Comment (zhenik) [inline template/dev/README.md:8]: Put `them`? What exactly to put?

File: template/test_example/dev/ansible/00_test_namespace_seperation.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  debug:
+    msg: asd
+  when: lookup( 'env', 'consul_acl' ) | bool and lookup( 'env', 'consul_enterprise' ) | bool
+  tags: test
+
+- name: some test should run
+  debug:
+    msg: asd
+  when: not lookup( 'env', 'nomad_acl' ) | bool
+  tags: test
```
Comment (pdmthorsrud) [inline template/test_example/dev/ansible/00_test_namespace_seperation.yml:15]:
```suggestion
```
I wasn't brave enough to click "Delete", but this should be removed, right?

File: template/test_example/dev/ansible/00_test_namespace_seperation.yml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+- name: Debug all env vars
+  debug:
+    var: ansible_env
+
+- name: some test should not run
```
Comment (zhenik) [inline template/test_example/dev/ansible/00_test_namespace_seperation.yml:5]: Is it example?

Comment (zhenik) [reply template/test_example/dev/ansible/00_test_namespace_seperation.yml:5]: duplicate of @pdmthorsrud question above

File: template/Makefile (hunk 2 above)
Comment (fredrikhgrelland) [reply template/Makefile:14]: https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables

File: template/test_example/dev/ansible/00_test_namespace_seperation.yml (hunk 4 above)
Comment (fredrikhgrelland) [reply template/test_example/dev/ansible/00_test_namespace_seperation.yml:15]: Yeah. We can delete it

File: template/Makefile (hunk 2 above)
Comment (pdmthorsrud) [reply template/Makefile:14]: Ah, sweet. Thanks

File: template/dev/README.md (hunk 3 above)
Comment (fredrikhgrelland) [reply template/dev/README.md]: This doc is all wrong.

File: ansible/bootstrap.yml [hunk 6]
```diff
@@ -8,25 +8,31 @@
       - nomad
     token: master
   tasks:
+    - name: Check if .env_override is provided
+      debug:
+        msg: "We are running in CI and .env_override is set"
+      failed_when: local_test is undefined and not lookup( 'env', 'TF_VAR_env_override') | bool
+      tags: test
+
```
Comment (pdmthorsrud) [inline ansible/bootstrap.yml:16]:
```suggestion
    - name: Check if .env_override is provided
      debug:
        msg: "Checking if env_override process was successful. Doing nothing if this is a local test"
      failed_when: local_test is undefined and not lookup( 'env', 'TF_VAR_env_override') | bool
      tags: test

```

Comment (pdmthorsrud) [reply ansible/bootstrap.yml:16]: Is this showing up as no changes for you as well?

Anyways, I've changed the message.

**Comment by zhenik** [comment]
@fredrikhgrelland I suggest to merge it as it is. And continue fixing in new small PR, due that current PR become non-reviewable with many changes and several changes' contexts

**Pull request merged**